src/mainloop.rkt

```diff
@@ -319,78 +318,16 @@
+(define (run-improve! expression context pcontext rules iterations)
   (timeline-event! 'preprocess)
-
-  ;; If the specification is given, it is used for sampling points
-  (define sortable (connected-components specification (*context*)))
-
-  (define new-preprocess
-    (for/list ([sortable-variables (in-list sortable)]
-               #:when (> (length sortable-variables) 1))
-      (cons 'sort sortable-variables)))
-  (timeline-push! 'symmetry (map ~a new-preprocess))
-  (*herbie-preprocess* (append preprocess new-preprocess))
-
-  (preprocess-pcontext pcontext (*herbie-preprocess*) (*context*)))
-
-
-(define (run-improve! prog pcontext iters
-                      #:specification [specification #f]
-                      #:preprocess [preprocess '()])
-  (define processed-pcontext
-    (make-preprocess-pcontext prog pcontext iters
-                              #:specification specification
-                              #:preprocess preprocess))
-
-  (match-define (cons best rest) (mutate! prog iters processed-pcontext))
-
-  (*herbie-preprocess* (remove-unnecessary-preprocessing best pcontext (*herbie-preprocess*)))
-  (cons best rest))
-
-(define (preprocessing-<=? alt pcontext preprocessing-one preprocessing-two ctx)
-  (define vars (context-vars ctx))
-  (define pcontext1 (preprocess-pcontext pcontext preprocessing-one ctx))
-  (define pcontext2 (preprocess-pcontext pcontext preprocessing-two ctx))
-  (<= (errors-score (errors (alt-expr alt) pcontext1 (*context*)))
-      (errors-score (errors (alt-expr alt) pcontext2 (*context*)))))
-
-(define (drop-at ls index)
-  (define-values (front back) (split-at ls index))
-  (append front (rest back)))
-
-; until fixed point, iterate through preprocessing attempting to drop preprocessing with no effect on error
-(define (remove-unnecessary-preprocessing alt pcontext preprocessing #:removed [removed empty])
-  (define-values (result newly-removed)
-    (let loop ([preprocessing preprocessing] [i 0] [removed removed])
-      (cond
-        [(>= i (length preprocessing))
-         (values preprocessing removed)]
-        [(preprocessing-<=? alt pcontext (drop-at preprocessing i) preprocessing (*context*))
-         (loop (drop-at preprocessing i) i (cons (list-ref preprocessing i) removed))]
-        [else
-         (loop preprocessing (+ i 1) removed)])))
-  (cond
-    [(< (length result) (length preprocessing))
-     (remove-unnecessary-preprocessing alt pcontext result #:removed newly-removed)]
-    [else
-     (timeline-push! 'remove-preprocessing (map ~a newly-removed))
-     result]))
+  (define preprocessing (find-preprocessing expression context rules))
+  (timeline-push! 'symmetry (map ~a preprocessing))
+  (define pcontext* (preprocess-pcontext context pcontext preprocessing))
+  (match-define (and alternatives (cons best _))
+    (mutate! expression iterations pcontext*))
+  (define preprocessing*
+    (remove-unnecessary-preprocessing (alt-expr best) context pcontext preprocessing))
```

Comment:
I think this really emphasizes the weirdness here, which is that we only look at the best expr when removing unnecessary preprocessing. It would probably be better if we did this separately for each alt, but right now there are plumbing problems preventing that. Anyway, just a thought.

Reply:
Yeah this was definitely a don't-totally-understand-this-just-copy-it-over part of the changes.

Cool! What kind of plumbing problems specifically? And if we did it per alt, would we have some way of attaching preprocessing steps to an alt (at least on the way out of Herbie)?

Reply:
Exactly: there's no way to attach preprocessing steps to an alt. Our output is just preprocessing + list of alts. I think what we need to do is add preprocessing into the alt, possibly by adding it to the AST itself using something like `(herbie-preprocess ...)` or even `(herbie-sort ...)`. Not sure, worth discussing in a meeting if we can grab a slot.

The issue with making it part of the AST is whether the rest of Herbie knows what to do with them. Not sure! I know the TeXifyer wouldn't work. If that was the only challenge, we could work around it, but there are probably others. I've been toying with the idea that an alt should store a preprocessing & an expr, which is nice and safe and would work with derivations. Anyway!